JV - Replace Paperclip with ActiveStorage

jbvanmeter opened the pull request:
Implements ActiveStorage to manage new model attachments, identifies/migrates all existing Paperclip attachment data required by Activestorage to the relevant ActiveStorage database tables/columns, and copies/migrates all attachment files in Paperclip directories to ActiveStorage directories. ActiveStorage is then fully managing all attachments, both those originally managed by Paperclip and any new attachments that will be created within SPARC. Drops database columns created by Paperclip and removes the gem.

https://www.pivotaltracker.com/story/show/170086933

[#170086933]

Commit:
After ActiveStorage install and configure

Commit:
Source code updates for ActiveStorage implementation

Commit:
Fixed issue with ActiveStorage blob records not being deleted and remaining orphaned after attachment deletion

Commit:
All pre-migration updates

Commit:
Successful transplant of relevant Paperclip data to ActiveStorage tables

Commit:
Successful copying of all Paperclip attachment files from original Paperclip directories to valid ActiveStorage directories.

Commit:
All Paperclip generated model column names have been dropped from the database.

Commit:
Remaining local debugging after completion of all migrations, removal of Paperclip gem, install of required image processing gems for ActiveStorage.

Commit:
ImageProcessing gem installed

Commit:
Gemfile.lock updated with ImageProcessing

amcates commented:
@jbvanmeter This now has merge conflicts.  Please resolve them and re-push

Commit:
Manually changed db/schema.rb to match what is currently on v3.10.0 to allow succesfful merging.

Commit:
Updated header version for db/schema.rb to match what is currently on v3.10.0

jbvanmeter commented:
db/schema.rb manually updated to match what is currently on v3.10.0

amcates commented on db/schema.rb:
```diff
@@ -1105,7 +1119,7 @@
   end
 
-  create_table "visit_groups", options: "ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci", force: :cascade do |t|
+  create_table "visit_groups", options: "ENGINE=InnoDB DEFAULT CHARSET=utf8mb3 COLLATE=utf8_unicode_ci", force: :cascade do |t|
```
Will the change to utf8mb3 cause a problem for future installs?  I ask because it won't change on our current deployment

jbvanmeter replied:
That shouldn't be a change resulting from any edits on this branch (I didn't make any edits to or associated with the 'visit_groups' table). I copied and pasted directly from the db/schema.rb on the v3.10.0 git repo.

amcates commented on db/schema.rb:
```diff
@@ -1117,13 +1131,13 @@
     t.datetime "updated_at", null: false
     t.datetime "deleted_at"
-    t.integer "research_billing_qty", default: 0
+    t.bigint "research_billing_qty", default: 0
```
Not sure why this one changed

jbvanmeter replied:
(Same as above, no changes made to or associated with 'research_billing_qty', db/schema.rb copied from git repo)

amcates commented:
@jbvanmeter Please resolve conflicts and then I will go a head and merge so that it doesn't have more conflicts.

jleonardw9 committed:
Resolving merge conflicts

jleonardw9 committed:
Delete delete_potential_funding_values.rake

This was an uncommitted change in my funding source branch when I switched to this one to resolve merge conflicts.

jleonardw9 reviewed (changes requested):
I've sent you a Teams message, but Travis is failing on an old migration creating documents. I think it's dependent on the paperclip gem (add_attachments is what it's complaining about). Everything else looks great.

Commit:
Add Paperclip back to gemfile so that previous migrations referencing Paperclip syntax can run successfully.

Commit:
Source code updates to payment_upload and report models, bug fix for new document creation.

Commit:
Fixing schema.rb merge conflict

Commit:
Bug fix attempt for Travis build, re: get_value NoMethodError

Commit:
Attempted bug fix for Travis build, Rspec edits

Commit:
Travis build error fix attempt

Commit:
Third attempted bug fix for Travis CI

Commit:
Replacing eager_load in migration to only query specific models - Document, PaymentUpload, and Report

Commit:
Eager load fix in migration 4 of 4

Commit:
Adding ActiveStorage config to config/environments/test.rb to fix Travis build error; additional source code housekeeping

Commit:
Documents helper spec file showed conflicts? Attempting fix.

Commit:
Why, God. Why.

Commit:
Updates to FactoryBot spec files for document.rb

Commit:
Syntax update for FactoryBot document.rb

Commit:
Rspec updates for payment_uploads and reports

Commit:
Attempted bug fix for MySQL connection disruption during Rspec ActiveStorage attachment test while Travis CI build runs

Commit:
Additional Rspec bug fix attempts for Travis CI

jbvanmeter committed:
Merge branch 'v3.10.0' into replace_paperclip_with_activestorage

Commit:
FactoryBot and model spec syntax updates, simplifications

Commit:
db/schema.rb alignment with v3.10.0

Commit:
rspec updates
